guppylang/decorator.py

```diff
@@ -241,11 +248,23 @@ def dec(c: type) -> GuppyDefinition:
             )
             mod.register_def(defn)
             mod._register_buffered_instance_funcs(defn)
-            return GuppyDefinition(defn)
+            # We're pretending to return the class unchanged, but in fact we return
+            # a `GuppyDefinition` that handles the comptime logic
+            return GuppyDefinition(defn)  # type: ignore[return-value]
 
         return dec
 
-    @property
+    @overload
+    @dataclass_transform()
```

Comment:
Mypy doesn't support `dataclass_transform` on methods yet: https://github.com/python/mypy/issues/16527

So this is only a benefit for pyright atm